### ISSUE TITLE
Error with Next.js when using deserify in a selector

### ISSUE THREAD

**Issue by tuffstuff9** [karmaniverous/serify-deserify#24]
Hi, I have followed the setup guide provided in the Redux documentation for Next.js (https://redux.js.org/usage/nextjs).

I am trying to deserialize certain fields of my order state directly within the `useAppSelector` selector function. However, when I do this, I encounter an error. Here are the three approaches I have tried:

1. Retrieving the order state without deserializing:
```javascript
const rowData = useAppSelector(selectOrders);
```
This approach works as expected, but it requires me to deserialize the necessary fields wherever I need to use them, which is not ideal.

2. Deserializing within the selector:
```javascript
const rowData = useAppSelector((state) => {
    return state.app.orders.map((order) => ({
        ...order,
        eta: deserify(order.eta, defaultOptions),
        etd: deserify(order.etd, defaultOptions),
        created_at: deserify(order.created_at, defaultOptions),
    }));
});
```
When I try to deserialize the fields directly within the selector, I encounter the following error:
```
Error
    at eval (webpack-internal:///(ssr)/./node_modules/react-redux/dist/react-redux.mjs:122:25)
    at memoizedSelector (webpack-internal:///(ssr)/./node_modules/use-sync-external-store/cjs/use-sync-external-store-with-selector.development.js:78:30)
    at getServerSnapshotWithSelector (webpack-internal:///(ssr)/./node_modules/use-sync-external-store/cjs/use-sync-external-store-with-selector.development.js:137:14)
    at Object.useSyncExternalStore (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:2797)
    at t.useSyncExternalStore (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:163011)
    at useSyncExternalStoreWithSelector (webpack-internal:///(ssr)/./node_modules/use-sync-external-store/cjs/use-sync-external-store-with-selector.development.js:144:15)
    at useSelector2 (webpack-internal:///(ssr)/./node_modules/react-redux/dist/react-redux.mjs:161:27)
    at GridExample (webpack-internal:///(ssr)/./src/app/(mainframe)/gridtwo/page.tsx:60:85)
    at au (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:10446)
    at /Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:15122
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16908)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at ab (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:11808)
    at /Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16871
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16908)
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:46505)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at aj (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:53793)
    at aC (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:52874)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:50095)
    at aj (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:53793)
    at aC (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:52874)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:50095)
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:45321)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at ab (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:11808)
    at /Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16871
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16908)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at ac (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:11400)
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:14707)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at ab (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:11808)
    at /Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16871
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16908)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:42922)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at ab (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:11808)
    at /Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16871
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16908)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:42922)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at ab (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:11808)
    at /Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16871
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:16908)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
    at aw (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:42922)
    at a_ (/Users/user/dev/vscode/redux_deserify/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:49776)
```

3. Deserializing the selected orders separately:
```javascript
const orders = useAppSelector(selectOrders);

const rowData = orders.map((order) => ({
    ...order,
    eta: deserify(order.eta, defaultOptions),
    etd: deserify(order.etd, defaultOptions),
    created_at: deserify(order.created_at, defaultOptions),
}));
```
This approach works fine, but it is not ideal as it requires an additional step to deserialize the data after selecting it from the Redux store.

I would like to know if there is a way to deserialize the necessary fields directly within the `useAppSelector` selector function without encountering the error mentioned above. Any guidance or suggestions on how to handle this situation would be greatly appreciated.

Steps to Reproduce:
1. Set up a Next.js project with Redux using the guide provided in the Redux documentation.
2. Create a Redux slice for managing the order state.
3. Attempt to deserialize the required fields within the `useAppSelector` selector function.
4. Observe the error mentioned above.

Please let me know if you need any further information or clarification regarding this issue. Thank you for your assistance!

**Comment by karmaniverous** [comment]
This is going to be a bad approach no matter what. 

Think about how a selector works. It's a memoized function, so that as long as the output remains the same, your component doesn't refresh. But in this case you are constructing a new object every time the selector runs, so that it will NEVER be the same, even if its content doesn't change! So your component is doomed to experience runaway rendering.

You can mitigate this with my [blockRefresh](https://github.com/karmaniverous/block-refresh) package. I haven't refactored that one to TS yet but [the logic](https://github.com/karmaniverous/block-refresh/blob/main/lib/blockRefresh/blockRefresh.js) is pretty simple so feel free to duplicate.

Now regarding the error in the selector... Redux is a little greedy around objects in its store, so when you hand them off to `deserify` inside the selector there may be trouble.

Just for fun, try converting bits like `deserify(order.eta, defaultOptions)` to `deserify(_.cloneDeep(order.eta), defaultOptions)`. If that solves the problem, then you're back in business and I'll think about how best to deal with that from inside.

**Comment by karmaniverous** [comment]
Another thought regarding your selectors... in a case like yours, you're probably going to be way better off with multiple selectors--say ne for each deserified value--rather than just one.

The reason goes back to memoization. Running an additional selector costs almost nothing, it's just accessing an object property in browser memory. But when the resulting value changes, it can trigger a whole tree of updates in dependent values (`useMemo` etc). 

Since your values are nicely segregated into their own properties, it makes sense to limit the blast radius of any one change by NOT combining them into a single selector.

**Comment by karmaniverous** [comment]
So @tuffstuff9 cloning on `deserify` is obviously the right thing to do... in fact I'm pretty sure I was doing that before the TS refactor, it just got away from me.

Anyway I've added it back, so if you update to the latest version it should solve the deserify-in-selector issue. Just please keep in mind my comments above.

Hm also... forget `deserify` for a moment. Say you pull a single object-type property out of your store. Unless you use `blockRefresh` or some other `equalityFn`, your component will NEVER rerender, for the same reason described above: the selector is a memoized function, and the thing being memoized is the object **identifier**, not its contents.

Bottom line: if you are storing objects in Redux (as opposed to primitive types) you MUST use the `equalityFn` selector param. More info [here](https://react-redux.js.org/api/hooks#useselector).

Anyway let me know if the update works for you.

**Comment by tuffstuff9** [comment]
Thanks for your thorough response! 

Unfortunately, the update did not get rid of the error. I believe it may have something to do with a discrepancy between the server state and client state (note the [comprehensive setup for use with next.js](https://redux.js.org/usage/nextjs), which sets it up specifically to work with server state - I believe it may have something to do with that, hence the `getServerSnapshotWithSelector` in the error.)

I am also confused about your comments. 

> But in this case you are constructing a new object every time the selector runs

Is this because we are deserifying into a Date, which is the 'new object'? Won't a selector return a new object anyway because of the map? I'm not sure I follow your thinking. The selectors are memoized, and only recalculate when the inputs change, right? So in that case, what is the difference between retrieving the order's dates normally (assuming dates were serializable) and deserifying it (given that dates are not _actually serializable_)? That leads me on to address another comment:

> in a case like yours, you're probably going to be way better off with multiple selectors

The issue is that I'm using ag-grid, and it expects a rowData object, which is pretty much my orders, with a bunch of deeply nested objects for sub grids (like `order_lines` for example), nested and all.

> so that it will NEVER be the same, even if its content doesn't change!

But getting back to this, if the content (i.e input) doesn't change, then the selector wouldn't even be run, right?

> Since your values are nicely segregated into their own properties, it makes sense to limit the blast radius of any one change by NOT combining them into a single selector.

So do you mean that instead of something like:

```tsx
		selectOrders: (state) => {
			return state.app.orders.map((order) => ({
				...order,
				eta: deserify(order.eta, defaultOptions),
				etd: deserify(order.etd, defaultOptions),
				created_at: deserify(order.created_at, defaultOptions),
			}));
		},
```

We would instead have multiple selectors where we pass a section of the required state into new selectors, (i.e passing in `orders` rather than `state`), and then use those in a new selector?

> cloning on `deserify` is obviously the right thing to do

I've spent quite a while trying to understand why we would need this. I think I'm probably missing something obvious. I've taken a look at the source code and there doesn't seem to be any mutations in either the `deserify` or `_deserify` functions. So what reason would we have for deep cloning?

> Unless you use `blockRefresh` or some other `equalityFn`

Since redux is immutable, won't it create a new object reference on every state change?

Sorry for the barrage of questions!

**Comment by karmaniverous** [comment]
Even if you just did this:

```ts
const order = useAppSelector((state) => {
    return state.app.orders.map((order) => ({ ...order }));
});
```

You would in fact be returning a new object each time your selector is called. Think about how equality works for JavaScript objects. The assertion `({} === {})` is ALWAYS `false`—try it!—because even though the two objects declared inline have the same CONTENT, they don't have the same REFERENCE.

`42 === 42`, on the other hand, is always `true`... because, in JavaScript, primitive values are always compared by VALUE whereas objects are always compared by REFERENCE.

When you apply your rest operator to a store object (i.e. `{ ...order }`) you are creating a new object whose VALUES are equal to the VALUES in the original object. If those values are primitives, they are equal by value. But if they are objects, then they are equal by REFERENCE, which is just another way of saying that the keys in your new object are pointing directly at objects in your store.

So if I do this (assuming `order` already has a `foo` property):

```ts
const order = useAppSelector((state) => {
    return state.app.orders.map((order) => ({ ...order }));
});

order.foo.bar = 'baz';
```

... then I am attempting a direct manipulation of an object inside my Redux store, which the Redux and Immer guys have taken great pains to prevent.

> The selectors are memoized, and only recalculate when the inputs change, right?

No. A Selector has no way of knowing if the value it is pulling out of the Store has changed unless it goes and looks. You will note that the is no dependency array on a Selector like there is on `useMemo`. So a Selector retrieves its output every time the component renders.

The question is what happens NEXT. The Selector doesn't always pass its output to the left side of the assignment. That only happens if the Selector's output has CHANGED, either with a new value (if it's a primitive) or a new reference (if it's an object). This is the change that triggers everything dependent on that value to update.

The job of the `equalityFn` argument is to exert manual control over this process, since it's obviously pointless to compare the references of objects constructed inside the selector... they change every time.

> Since redux is immutable, won't it create a new object reference on every state change?

Hm maybe (although that's kind of a sweeping statement)... but it doesn't matter in any case. You need to control refresh on the basis of your object's VALUES, not its reference. 

> ... the right thing to do?

You know you are absolutely right. When I refactored for TS I moved the (quite necessary) cloning process [inside the `deserify` function](https://github.com/karmaniverous/serify-deserify/blob/c9c62ccc541d80af774a74b37f0d5fe078c00e54/src/deserify/deserify.ts#L25-L29). Then when I looked at it the other day, I just didn't see it. The new extra cloning up front imposes a minor but completely unnecessary performance hit, and I'll remove it in the next release. 

One last thing... there should be no need for you to deserify your order object by parts. Just deserify the whole thing, and any serified object inside will be deserified recursively. So:

```ts
const rowData = useAppSelector((state) => deserify(state.app.orders, defaultOptions));
```

... or if that doesn't work:

```ts
const orders = useAppSelector(selectOrders);
const rowData = deserify(orders, defaultOptions);
```

... only you're writing a React component, so:

```ts
const orders = useAppSelector(selectOrders);
const rowData = useMemo(() => deserify(orders, defaultOptions), [orders]);
```

**Comment by tuffstuff9** [comment]
Thanks for the detailed reply - you helped to clear a few things up for sure.

> ... only you're writing a React component, so:
> 
> ```ts
> const orders = useAppSelector(selectOrders);
> const rowData = useMemo(() => deserify(orders, defaultOptions), [orders]);
> ```

This seems to have done the trick! I hadn't realised we could just wrap the entire selector with deserify.
